aiida/backends/sqlalchemy/migrations/utils/legacy_workflows.py

```diff
@@ -0,0 +1,76 @@
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+# pylint: disable=invalid-name
+"""Utilities for removing legacy workflows."""
+import sys
+
+import click
+from sqlalchemy.sql import func, select, table
+
+from aiida.cmdline.utils import echo
+from aiida.common import json
+
+
+def json_serializer(obj):
+    """JSON serializer for objects not serializable by default json code"""
+    from datetime import date, datetime
+    from uuid import UUID
+
+    if isinstance(obj, UUID):
+        return str(obj)
+
+    if isinstance(obj, (datetime, date)):
+        return obj.isoformat()
+
+    raise TypeError(f'Type {type(obj)} not serializable')
+
+
+def export_workflow_data(connection, profile):
+    """Export existing legacy workflow data to a JSON file."""
+    from tempfile import NamedTemporaryFile
+
+    DbWorkflow = table('db_dbworkflow')
+    DbWorkflowData = table('db_dbworkflowdata')
+    DbWorkflowStep = table('db_dbworkflowstep')
+
+    count_workflow = connection.execute(select(func.count()).select_from(DbWorkflow)).scalar()
+    count_workflow_data = connection.execute(select(func.count()).select_from(DbWorkflowData)).scalar()
+    count_workflow_step = connection.execute(select(func.count()).select_from(DbWorkflowStep)).scalar()
+
+    # Nothing to do if all tables are empty
+    if count_workflow == 0 and count_workflow_data == 0 and count_workflow_step == 0:
+        return
+
+    if not profile.is_test_profile:
+        echo.echo('\n')
+        echo.echo_warning('The legacy workflow tables contain data but will have to be dropped to continue.')
+        echo.echo_warning('If you continue, the content will be dumped to a JSON file, before dropping the tables.')
+        echo.echo_warning('This serves merely as a reference and cannot be used to restore the database.')
+        echo.echo_warning('If you want a proper backup, make sure to dump the full database and backup your repository')
+        if not click.confirm('Are you sure you want to continue', default=True):
+            sys.exit(1)
+
+    delete_on_close = profile.is_test_profile
+
+    data = {
+        'workflow': [dict(row) for row in connection.execute(select('*').select_from(DbWorkflow))],
+        'workflow_data': [dict(row) for row in connection.execute(select('*').select_from(DbWorkflowData))],
+        'workflow_step': [dict(row) for row in connection.execute(select('*').select_from(DbWorkflowStep))],
+    }
+
+    with NamedTemporaryFile(
+        prefix='legacy-workflows', suffix='.json', dir='.', delete=delete_on_close, mode='w+'
```

Comment:
This was causing migration to fail because the data returned from database are bytes.
```suggestion
        prefix='legacy-workflows', suffix='.json', dir='.', delete=delete_on_close, mode='wb'
```

Reply:
cheers, added this in [e84f38c](https://github.com/aiidateam/aiida-core/pull/5330/commits/e84f38cf64c3dbcb6b284d95af43fd61d35d5623), with a test

At a guess, this has been broken since #2107 😬

Reply:
Don't think so, the correct mode is being used in `develop`. You changed it to `w+` for some reason. The migration ran just fine on the `develop` branch. It is still running (now at the repo migration). Will report back later when it is done. Think with this fix, all that remains is essentially sorting out the naming of all the Django indices and constraints.

Reply:
Ah yeh, it was that the sqlalchemy migration had `w+`, i.e. was broken (aiida/backends/sqlalchemy/migrations/versions/1b8ed3425af9_remove_legacy_workflows.py), but the django one had `wb`

Reply:
> The migration ran just fine on the develop branch.

When you say fine, it would be very helpful, if you could create a YAML, using the code in `tests/backends/aiida_sqlalchemy/migrations/conftest.py::schema_generator`, for both the initial database and final one after the `develop` migration. So we can actually see what these schemas are

Reply:
(in fact, it would be helpful to have them for every migration step, if you had the time an inclination to figure out how to do that lol)

Reply:
Added the original and migrated schemas. Let's start with that and see if we need more.

Reply:
done